### PR TITLE
moved to new jsonpath library

### PR DESCRIPTION
**Pull request by MeirShpilraien**
The new library provides a better performance when there is a need to also return the json path that led to a specific result.
This property allow to improve the write performance (and is some cases also the read perforce). The improvements are showed in the following table:

|                           Test Case                            |Baseline master (median obs. +- std.dev)|Comparison feature-new-jsonpath-lib (median obs. +- std.dev)|% change (higher-better)|        Note         |
|----------------------------------------------------------------|----------------------------------------|------------------------------------------------------------|------------------------|---------------------|
|json_arrappend_geojson                                          | 61527 +- nan%  (1 datapoints)          | 70161 +- nan%  (1 datapoints)                              |14.0%                   |IMPROVEMENT          |
|json_get_ResultSet.totalResultsAvailable_jsonsl-yahoo2_json     | 133326 +- nan%  (1 datapoints)         | 137923 +- nan%  (1 datapoints)                             |3.4%                    |potential IMPROVEMENT|
|json_get_[0]_jsonsl-1                                           | 136043 +- nan%  (1 datapoints)         | 147050 +- nan%  (1 datapoints)                             |8.1%                    |IMPROVEMENT          |
|json_get_[7]_jsonsl-1                                           | 145977 +- nan%  (1 datapoints)         | 150367 +- nan%  (1 datapoints)                             |3.0%                    |potential IMPROVEMENT|
|json_get_[8].zero_jsonsl-1                                      | 136043 +- nan%  (1 datapoints)         | 139852 +- nan%  (1 datapoints)                             |2.8%                    |-- no change --      |
|json_get_[web-app].servlet[0][servlet-name]_json-parser-0000    | 119389 +- nan%  (1 datapoints)         | 124992 +- nan%  (1 datapoints)                             |4.7%                    |potential IMPROVEMENT|
|json_get_[web-app].servlet[0]_json-parser-0000                  | 90080 +- nan%  (1 datapoints)          | 94778 +- nan%  (1 datapoints)                              |5.2%                    |IMPROVEMENT          |
|json_get_[web-app].servlet_json-parser-0000                     | 81623 +- nan%  (1 datapoints)          | 82978 +- nan%  (1 datapoints)                              |1.7%                    |-- no change --      |
|json_get_array_of_docs[1]_pass_100_json                         | 129022 +- nan%  (1 datapoints)         | 134221 +- nan%  (1 datapoints)                             |4.0%                    |potential IMPROVEMENT|
|json_get_array_of_docs[1]sclr_pass_100_json                     | 130709 +- nan%  (1 datapoints)         | 133323 +- nan%  (1 datapoints)                             |2.0%                    |-- no change --      |
|json_get_array_of_docs_pass_100_json                            | 132440 +- nan%  (1 datapoints)         | 121942 +- nan%  (1 datapoints)                             |-7.9%                   |REGRESSION           |
|json_get_fulldoc_json-parser-0000                               | 87712 +- nan%  (1 datapoints)          | 84736 +- nan%  (1 datapoints)                              |-3.4%                   |potential REGRESSION |
|json_get_fulldoc_jsonsl-1                                       | 102556 +- nan%  (1 datapoints)         | 101000 +- nan%  (1 datapoints)                             |-1.5%                   |-- no change --      |
|json_get_fulldoc_jsonsl-yahoo2_json                             | 26838 +- nan%  (1 datapoints)          | 24992 +- nan%  (1 datapoints)                              |-6.9%                   |REGRESSION           |
|json_get_fulldoc_jsonsl-yelp_json                               | 13504 +- nan%  (1 datapoints)          | 13324 +- nan%  (1 datapoints)                              |-1.3%                   |-- no change --      |
|json_get_fulldoc_pass_100_json                                  | 130709 +- nan%  (1 datapoints)         | 129022 +- nan%  (1 datapoints)                             |-1.3%                   |-- no change --      |
|json_get_key_empty                                              | 161280 +- nan%  (1 datapoints)         | 149245 +- nan%  (1 datapoints)                             |-7.5%                   |REGRESSION           |
|json_get_message.code_jsonsl-yelp_json                          | 141832 +- nan%  (1 datapoints)         | 139852 +- nan%  (1 datapoints)                             |-1.4%                   |-- no change --      |
|json_get_sclr_pass_100_json                                     | 149249 +- nan%  (1 datapoints)         | 139848 +- nan%  (1 datapoints)                             |-6.3%                   |REGRESSION           |
|json_get_sub_doc.sclr_pass_100_json                             | 140837 +- nan%  (1 datapoints)         | 140837 +- nan%  (1 datapoints)                             |0.0%                    |                     |
|json_get_sub_doc_pass_100_json                                  | 137920 +- nan%  (1 datapoints)         | 132440 +- nan%  (1 datapoints)                             |-4.0%                   |potential REGRESSION |
|json_numincrby_num_1                                            | 136047 +- nan%  (1 datapoints)         | 140837 +- nan%  (1 datapoints)                             |3.5%                    |potential IMPROVEMENT|
|json_nummultby_num_2                                            | 142849 +- nan%  (1 datapoints)         | 142849 +- nan%  (1 datapoints)                             |0.0%                    |                     |
|json_set_ResultSet.totalResultsAvailable_1_jsonsl-yahoo2_json   | 117639 +- nan%  (1 datapoints)         | 130709 +- nan%  (1 datapoints)                             |11.1%                   |IMPROVEMENT          |
|json_set_[0]foo_jsonsl-1                                        | 125777 +- nan%  (1 datapoints)         | 125777 +- nan%  (1 datapoints)                             |0.0%                    |                     |
|json_set_[web-app].servlet[0][servlet-name]_bar_json-parser-0000| 90488 +- nan%  (1 datapoints)          | 107518 +- nan%  (1 datapoints)                             |18.8%                   |IMPROVEMENT          |
|json_set_fulldoc_api_replies_q1_google_autocomplete             | 42587 +- nan%  (1 datapoints)          | 43578 +- nan%  (1 datapoints)                              |2.3%                    |-- no change --      |
|json_set_fulldoc_api_replies_q2_gmaps_areatraffic               | 75892 +- nan%  (1 datapoints)          | 77962 +- nan%  (1 datapoints)                              |2.7%                    |-- no change --      |
|json_set_fulldoc_api_replies_q3_gmaps_passiveassist             | 38487 +- nan%  (1 datapoints)          | 39319 +- nan%  (1 datapoints)                              |2.2%                    |-- no change --      |
|json_set_fulldoc_api_replies_q4_gmaps_assist                    | 97551 +- nan%  (1 datapoints)          | 99742 +- nan%  (1 datapoints)                              |2.2%                    |-- no change --      |
|json_set_fulldoc_api_replies_q5_gmaps_place                     | 1447 +- nan%  (1 datapoints)           | 1494 +- nan%  (1 datapoints)                               |3.3%                    |potential IMPROVEMENT|
|json_set_fulldoc_pass-json-parser-0000                          | 31335 +- nan%  (1 datapoints)          | 31633 +- nan%  (1 datapoints)                              |1.0%                    |-- no change --      |
|json_set_fulldoc_pass_100_json                                  | 81622 +- nan%  (1 datapoints)          | 84024 +- nan%  (1 datapoints)                              |2.9%                    |-- no change --      |
|json_set_fulldoc_yahoo2                                         | 7906 +- nan%  (1 datapoints)           | 8033 +- nan%  (1 datapoints)                               |1.6%                    |-- no change --      |
|json_set_key_empty                                              | 152662 +- nan%  (1 datapoints)         | 144919 +- nan%  (1 datapoints)                             |-5.1%                   |REGRESSION           |
|json_set_message.code_1_jsonsl-yelp_json                        | 112349 +- nan%  (1 datapoints)         | 133326 +- nan%  (1 datapoints)                             |18.7%                   |IMPROVEMENT          |
|json_set_num_0                                                  | 151506 +- nan%  (1 datapoints)         | 162591 +- nan%  (1 datapoints)                             |7.3%                    |IMPROVEMENT          |
|json_set_sclr_1_pass_100_json                                   | 121203 +- nan%  (1 datapoints)         | 139852 +- nan%  (1 datapoints)                             |15.4%                   |IMPROVEMENT          |
|json_set_sclr_pass_100_json                                     | 129026 +- nan%  (1 datapoints)         | 136975 +- nan%  (1 datapoints)                             |6.2%                    |IMPROVEMENT          |
|json_vs_hashes_hset_key_simple                                  | 170204 +- nan%  (1 datapoints)         | 170931 +- nan%  (1 datapoints)                             |0.4%                    |-- no change --      |
|json_vs_hashes_json.set_key_simple                              | 129024 +- nan%  (1 datapoints)         | 127787 +- nan%  (1 datapoints)                             |-1.0%                   |-- no change --      |

### CODE REVIEW

**Comment by MeirShpilraien** [comment]
### Comparison between master and feature-new-jsonpath-lib for metric: (OverallRates.overallOpsRate,Tests.OVERALL.Throughput_ops_sec_,Ops/sec,rps,Tests.Overall.rps). Time Period from a day ago to a month from now. (environment used: oss-standalone)
|                           Test Case                            |Baseline master (median obs. +- std.dev)|Comparison feature-new-jsonpath-lib (median obs. +- std.dev)|% change (higher-better)|        Note         |
|----------------------------------------------------------------|----------------------------------------|------------------------------------------------------------|------------------------|---------------------|
|json_arrappend_geojson                                          | 61527 +- nan%  (1 datapoints)          | 70161 +- nan%  (1 datapoints)                              |14.0%                   |IMPROVEMENT          |
|json_get_ResultSet.totalResultsAvailable_jsonsl-yahoo2_json     | 133326 +- nan%  (1 datapoints)         | 137923 +- nan%  (1 datapoints)                             |3.4%                    |potential IMPROVEMENT|
|json_get_[0]_jsonsl-1                                           | 136043 +- nan%  (1 datapoints)         | 147050 +- nan%  (1 datapoints)                             |8.1%                    |IMPROVEMENT          |
|json_get_[7]_jsonsl-1                                           | 145977 +- nan%  (1 datapoints)         | 150367 +- nan%  (1 datapoints)                             |3.0%                    |potential IMPROVEMENT|
|json_get_[8].zero_jsonsl-1                                      | 136043 +- nan%  (1 datapoints)         | 139852 +- nan%  (1 datapoints)                             |2.8%                    |-- no change --      |
|json_get_[web-app].servlet[0][servlet-name]_json-parser-0000    | 119389 +- nan%  (1 datapoints)         | 124992 +- nan%  (1 datapoints)                             |4.7%                    |potential IMPROVEMENT|
|json_get_[web-app].servlet[0]_json-parser-0000                  | 90080 +- nan%  (1 datapoints)          | 94778 +- nan%  (1 datapoints)                              |5.2%                    |IMPROVEMENT          |
|json_get_[web-app].servlet_json-parser-0000                     | 81623 +- nan%  (1 datapoints)          | 82978 +- nan%  (1 datapoints)                              |1.7%                    |-- no change --      |
|json_get_array_of_docs[1]_pass_100_json                         | 129022 +- nan%  (1 datapoints)         | 134221 +- nan%  (1 datapoints)                             |4.0%                    |potential IMPROVEMENT|
|json_get_array_of_docs[1]sclr_pass_100_json                     | 130709 +- nan%  (1 datapoints)         | 133323 +- nan%  (1 datapoints)                             |2.0%                    |-- no change --      |
|json_get_array_of_docs_pass_100_json                            | 132440 +- nan%  (1 datapoints)         | 121942 +- nan%  (1 datapoints)                             |-7.9%                   |REGRESSION           |
|json_get_fulldoc_json-parser-0000                               | 87712 +- nan%  (1 datapoints)          | 84736 +- nan%  (1 datapoints)                              |-3.4%                   |potential REGRESSION |
|json_get_fulldoc_jsonsl-1                                       | 102556 +- nan%  (1 datapoints)         | 101000 +- nan%  (1 datapoints)                             |-1.5%                   |-- no change --      |
|json_get_fulldoc_jsonsl-yahoo2_json                             | 26838 +- nan%  (1 datapoints)          | 24992 +- nan%  (1 datapoints)                              |-6.9%                   |REGRESSION           |
|json_get_fulldoc_jsonsl-yelp_json                               | 13504 +- nan%  (1 datapoints)          | 13324 +- nan%  (1 datapoints)                              |-1.3%                   |-- no change --      |
|json_get_fulldoc_pass_100_json                                  | 130709 +- nan%  (1 datapoints)         | 129022 +- nan%  (1 datapoints)                             |-1.3%                   |-- no change --      |
|json_get_key_empty                                              | 161280 +- nan%  (1 datapoints)         | 149245 +- nan%  (1 datapoints)                             |-7.5%                   |REGRESSION           |
|json_get_message.code_jsonsl-yelp_json                          | 141832 +- nan%  (1 datapoints)         | 139852 +- nan%  (1 datapoints)                             |-1.4%                   |-- no change --      |
|json_get_sclr_pass_100_json                                     | 149249 +- nan%  (1 datapoints)         | 139848 +- nan%  (1 datapoints)                             |-6.3%                   |REGRESSION           |
|json_get_sub_doc.sclr_pass_100_json                             | 140837 +- nan%  (1 datapoints)         | 140837 +- nan%  (1 datapoints)                             |0.0%                    |                     |
|json_get_sub_doc_pass_100_json                                  | 137920 +- nan%  (1 datapoints)         | 132440 +- nan%  (1 datapoints)                             |-4.0%                   |potential REGRESSION |
|json_numincrby_num_1                                            | 136047 +- nan%  (1 datapoints)         | 140837 +- nan%  (1 datapoints)                             |3.5%                    |potential IMPROVEMENT|
|json_nummultby_num_2                                            | 142849 +- nan%  (1 datapoints)         | 142849 +- nan%  (1 datapoints)                             |0.0%                    |                     |
|json_set_ResultSet.totalResultsAvailable_1_jsonsl-yahoo2_json   | 117639 +- nan%  (1 datapoints)         | 130709 +- nan%  (1 datapoints)                             |11.1%                   |IMPROVEMENT          |
|json_set_[0]foo_jsonsl-1                                        | 125777 +- nan%  (1 datapoints)         | 125777 +- nan%  (1 datapoints)                             |0.0%                    |                     |
|json_set_[web-app].servlet[0][servlet-name]_bar_json-parser-0000| 90488 +- nan%  (1 datapoints)          | 107518 +- nan%  (1 datapoints)                             |18.8%                   |IMPROVEMENT          |
|json_set_fulldoc_api_replies_q1_google_autocomplete             | 42587 +- nan%  (1 datapoints)          | 43578 +- nan%  (1 datapoints)                              |2.3%                    |-- no change --      |
|json_set_fulldoc_api_replies_q2_gmaps_areatraffic               | 75892 +- nan%  (1 datapoints)          | 77962 +- nan%  (1 datapoints)                              |2.7%                    |-- no change --      |
|json_set_fulldoc_api_replies_q3_gmaps_passiveassist             | 38487 +- nan%  (1 datapoints)          | 39319 +- nan%  (1 datapoints)                              |2.2%                    |-- no change --      |
|json_set_fulldoc_api_replies_q4_gmaps_assist                    | 97551 +- nan%  (1 datapoints)          | 99742 +- nan%  (1 datapoints)                              |2.2%                    |-- no change --      |
|json_set_fulldoc_api_replies_q5_gmaps_place                     | 1447 +- nan%  (1 datapoints)           | 1494 +- nan%  (1 datapoints)                               |3.3%                    |potential IMPROVEMENT|
|json_set_fulldoc_pass-json-parser-0000                          | 31335 +- nan%  (1 datapoints)          | 31633 +- nan%  (1 datapoints)                              |1.0%                    |-- no change --      |
|json_set_fulldoc_pass_100_json                                  | 81622 +- nan%  (1 datapoints)          | 84024 +- nan%  (1 datapoints)                              |2.9%                    |-- no change --      |
|json_set_fulldoc_yahoo2                                         | 7906 +- nan%  (1 datapoints)           | 8033 +- nan%  (1 datapoints)                               |1.6%                    |-- no change --      |
|json_set_key_empty                                              | 152662 +- nan%  (1 datapoints)         | 144919 +- nan%  (1 datapoints)                             |-5.1%                   |REGRESSION           |
|json_set_message.code_1_jsonsl-yelp_json                        | 112349 +- nan%  (1 datapoints)         | 133326 +- nan%  (1 datapoints)                             |18.7%                   |IMPROVEMENT          |
|json_set_num_0                                                  | 151506 +- nan%  (1 datapoints)         | 162591 +- nan%  (1 datapoints)                             |7.3%                    |IMPROVEMENT          |
|json_set_sclr_1_pass_100_json                                   | 121203 +- nan%  (1 datapoints)         | 139852 +- nan%  (1 datapoints)                             |15.4%                   |IMPROVEMENT          |
|json_set_sclr_pass_100_json                                     | 129026 +- nan%  (1 datapoints)         | 136975 +- nan%  (1 datapoints)                             |6.2%                    |IMPROVEMENT          |
|json_vs_hashes_hset_key_simple                                  | 170204 +- nan%  (1 datapoints)         | 170931 +- nan%  (1 datapoints)                             |0.4%                    |-- no change --      |
|json_vs_hashes_json.set_key_simple                              | 129024 +- nan%  (1 datapoints)         | 127787 +- nan%  (1 datapoints)                             |-1.0%                   |-- no change --      |

**Comment by gkorland** [comment]
@MeirShpilraien why do we still have `jsonpath_lib` in the code?

**Comment by gkorland** [comment]
ref #674

**Comment by gkorland** [comment]
ref #614

**Comment by MeirShpilraien** [comment]
> @MeirShpilraien why do we still have `jsonpath_lib` in the code?

https://github.com/RedisJSON/RedisJSON/blob/4fdbb23311763b9dc99a9a1fcea13ada82c58137/src/nodevisitor.rs#L1
Node visitor is still using it, I need to see the functionality it needs and find another way to do it...

File: tests/pytest/test_multi.py [hunk 1]
```diff
@@ -632,7 +634,8 @@ def testObjKeysCommand(env):
     r.assertOk(r.execute_command('JSON.SET', 'doc1', '$', '{"nested1": {"a": {"foo": 10, "bar": 20}}, "a":["foo"], "nested2": {"a": {"baz":50}}}'))
     # Test multi
     res = r.execute_command('JSON.OBJKEYS', 'doc1', '$..a')
-    r.assertEqual(res, [["foo", "bar"], None, ["baz"]])
+    res = reduce(lambda x, y: x + y if type(y) == list else x+[y], res, [])
+    r.assertEqual(set(res), set(["foo", "bar", None, "baz"]))
```
Comment (oshadmi) [inline tests/pytest/test_multi.py:638]: Change of order is a breaking change? Need to add label `pr:break`?

File: tests/pytest/test_multi.py [hunk 2]
```diff
@@ -663,7 +666,7 @@ def testObjLenCommand(env):
     r.assertOk(r.execute_command('JSON.SET', 'doc1', '$', '{"nested1": {"a": {"foo": 10, "bar": 20}}, "a":["foo"], "nested2": {"a": {"baz":50}}}'))
     # Test multi
     res = r.execute_command('JSON.OBJLEN', 'doc1', '$..a')
-    r.assertEqual(res, [2, None, 1])
+    r.assertEqual(set(res), set([2, None, 1]))
```
Comment (oshadmi) [inline tests/pytest/test_multi.py:669]: Same comment on whether a change of order is considered a breaking change

File: tests/pytest/test.py [hunk 3]
```diff
@@ -1000,7 +1000,8 @@ def testIssue_74(env):
 
     r.assertOk(r.execute_command('JSON.SET', 'test', '.', '{}'))
     # This shouldn't crash Redis
-    r.expect('JSON.SET', 'test', '$a', '12').raiseError()
+    r.expect('JSON.SET', 'test', '$a', '12').equal("OK")
```
Comment (oshadmi) [inline tests/pytest/test.py:1003]: This is a progression (bug fix) - plain `12` is now parsed properly as numeric, which is good.
Seems we need to document this (in related Release Notes)

File: src/commands.rs [hunk 4]
```diff
@@ -147,11 +151,14 @@ impl<'a, V: SelectValue> KeyValue<'a, V> {
...
 
     fn get_values<'b>(&'a self, path: &'b str) -> Result<Vec<&'a V>, Error> {
-        let mut selector = Selector::new();
-        selector.str_path(path)?;
-        selector.value(self.val);
-        let results = selector.select()?;
+        let query = compile(path)?;
+        let results = calc_once(query, self.val);
         Ok(results)
+        // let mut selector = Selector::new();
```
Comment (oshadmi) [inline src/commands.rs:157]: Can drop commented-out code (also in other locations)

File: src/commands.rs [hunk 5]
```diff
@@ -239,63 +252,62 @@ impl<'a, V: SelectValue> KeyValue<'a, V> {
...
-                }
-                selector.value(self.val);
-                let res = selector.select_with_paths(|_| true)?;
-                Ok(res
-                    .into_iter()
-                    .map(|v| {
-                        UpdateInfo::AUI(AddUpdateInfo {
-                            path: v,
-                            key: key.to_string(),
+        // let mut parsed_static_path = StaticPathParser::check(path)?;
```
Comment (oshadmi) [inline src/commands.rs:266]: More commented-out code

File: tests/pytest/test.py [hunk 6]
```diff
@@ -251,10 +251,10 @@ def testGetWithPathErrors(env):
     # If paths contain illegal characters, the error message must not contain them
 
     # Path (and error message) with embedded nulls in path
-    r.expect('JSON.GET', 'x', 'gar\x00\x00bage').raiseError().contains("does not exist")
+    r.expect('JSON.GET', 'x', 'gar\x00\x00bage').raiseError()
```
Comment (oshadmi) [inline tests/pytest/test.py:254]: Why not test also some of the error message?

File: tests/pytest/test.py [hunk 7]
```diff
@@ -835,7 +835,7 @@ def testLenCommands(env):
     r.expect('JSON.ARRLEN', 'test', '.arr[999]').raiseError().contains("does not exist")
 
     # test an infinite index
-    r.expect('JSON.ARRLEN', 'test', '.arr[-inf]').raiseError().contains("path error")
+    r.expect('JSON.ARRLEN', 'test', '.arr[-inf]').raiseError().contains("Error accured")
```
Comment (oshadmi) [inline tests/pytest/test.py:838]: Why not test also some of the error message details? (e.g., that it is a path related error)

**Review by oshadmi — COMMENTED** [review]
👍🏼 Only a few minor comments

File: tests/pytest/test_multi.py (hunk 1 above)
Comment (MeirShpilraien) [reply tests/pytest/test_multi.py:638]: Do we have such label?

File: tests/pytest/test_multi.py (hunk 2 above)
Comment (MeirShpilraien) [reply tests/pytest/test_multi.py:669]: I do not think it is considered a breaking change

File: tests/pytest/test.py (hunk 7 above)
Comment (MeirShpilraien) [reply tests/pytest/test.py:840]: It is a detailed error that explain what it is expect to get. I believe this is good enough.

File: tests/pytest/test_multi.py (hunk 2 above)
Comment (MeirShpilraien) [reply tests/pytest/test_multi.py:684]: @gkorland @K-Jo please approve/disapprove.

File: src/jsonpath/json_path.rs [hunk 8]
```diff
@@ -0,0 +1,1064 @@
...
+        self.is_static = Some(is_static);
+        self.is_static()
+    }
+}
+
+impl std::fmt::Display for QueryCompilationError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        write!(
+            f,
+            "Error accured on possition {}, {}",
```
Comment (oshadmi) [inline src/jsonpath/json_path.rs:112]:
```suggestion
            "Error occurred on position {}, {}",
```

File: src/jsonpath/json_path.rs [hunk 9]
```diff
@@ -0,0 +1,1064 @@
...
+                let start = 0;
+                let end =
+                    self.calc_abs_index(curr.next().unwrap().as_str().parse::<i64>().unwrap(), n);
+                let step = match curr.next() {
+                    Some(s) => s.as_str().parse::<usize>().unwrap(),
+                    None => 1,
+                };
+                (start, end, step)
+            }
+            Rule::all_range => {
```
Comment (oshadmi) [inline src/jsonpath/json_path.rs:711]: Seems that tests are not considered by codecov? (`array_range_step_all` should cover this, no?)

**Review by oshadmi — COMMENTED** [review]
Small change + comment

File: src/jsonpath/json_path.rs (hunk 9 above)
Comment (MeirShpilraien) [reply src/jsonpath/json_path.rs:711]: So this is a build system issue no?

File: tests/pytest/test_multi.py (hunk 2 above)
Comment (MeirShpilraien) [reply tests/pytest/test_multi.py:684]: The PR returns `[None, 2, 1]` which is also what is been returned here: https://jsonpath.com/

File: src/commands.rs [hunk 10]
```diff
@@ -239,63 +244,53 @@ impl<'a, V: SelectValue> KeyValue<'a, V> {
...
+        match token_type {
+            JsonPathToken::String => {
+                if query.size() == 1 {
+                    // Adding to the root
+                    Ok(vec![UpdateInfo::AUI(AddUpdateInfo {
+                        path: Vec::new(),
+                        key: last,
+                    })])
+                } else {
+                    // Adding somewhere in existing object, use jsonpath_lib::replace_with
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply src/commands.rs]: Fixed

File: src/commands.rs (hunk 4 above)
Comment (MeirShpilraien) [reply src/commands.rs]: Dropped

File: tests/pytest/test.py (hunk 6 above)
Comment (MeirShpilraien) [reply tests/pytest/test.py]: Fixed

File: tests/pytest/test_multi.py (hunk 2 above)
Comment (MeirShpilraien) [reply tests/pytest/test_multi.py:684]: @gkorland agrees its not a breaking change.

**Review by oshadmi — APPROVED** [review]
🏆

**Pull request merged**